lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,891 @@
+            break
+
+        status = resp.get('status', 'waiting')
+        display.vvv('Galaxy import process has a status of %s, wait %d seconds before trying again' % (status, wait))
+        time.sleep(wait)
+
+    for message in resp.get('messages', []):
+        level = message['level']
+        if level == 'error':
+            display.error("Galaxy import error message: %s" % message['message'])
+        elif level == 'warning':
+            display.warning("Galaxy import warning message: %s" % message['message'])
+        else:
+            display.vvv("Galaxy import message: %s - %s" % (level, message['message']))
+
+    if resp['state'] == 'failed':
+        code = resp['error'].get('code', 'UNKNOWN')
+        description = resp['error'].get('description', "Unknown error, see %s for more details" % task_url)
+        raise AnsibleError("Galaxy import process failed: %s (Code: %s)" % (description, code))
+
+
+def _find_existing_collections(path):
+    collections = []
+
+    for namespace in os.listdir(path):
+        namespace_path = os.path.join(path, namespace)
+        if os.path.isfile(namespace_path):
+            continue
+
+        for collection in os.listdir(namespace_path):
+            collection_path = os.path.join(namespace_path, collection)
+            if os.path.isdir(collection_path):
+                req = CollectionRequirement.from_path(collection_path, True)
+                display.vvv("Found installed collection %s:%s at '%s'" % (str(req), req.latest_version,
+                                                                          collection_path))
+                collections.append(req)
+
+    return collections
+
+
+def _build_dependency_map(collections, existing_collections, temp_path, servers, validate_certs, force, force_deps,
+                          no_deps):
+    dependency_map = {}
+
+    # First build the dependency map on the actual requirements
+    for name, version, source in collections:
+        _get_collection_info(dependency_map, existing_collections, name, version, source, temp_path, servers,
+                             validate_certs, (force or force_deps))
+
+    # Now parse the dependency requirements if no_deps was not set
+    if not no_deps:
+        checked_parents = set([str(c) for c in dependency_map.values() if c.skip])
+
+        while len(dependency_map) != len(checked_parents):
+            while True:
+                parents_to_check = set(dependency_map.keys()).difference(checked_parents)
+
+                deps_exhausted = True
+                for parent in parents_to_check:
+                    parent_info = dependency_map[parent]
+
+                    if parent_info.dependencies:
+                        deps_exhausted = False
+                        for dep_name, dep_requirement in parent_info.dependencies.items():
+                            _get_collection_info(dependency_map, existing_collections, dep_name, dep_requirement,
```

Comment:
This is on the 7th level of indentation and is hard to read.
Can you split subroutines a bit?